RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc

```diff
@@ -90,4 +103,133 @@ void PatternRecognitionbyCA::makeTracksters(const edm::Event &ev,
+                                                   std::vector<Trackster> &tracksters) {
+  // Energy regression and particle identification strategy:
+  // 1. Set default values for regressed energy and particle id for each trackster.
+  // 2. Store indices of tracksters whose total sum of cluster energies is above the
+  //    eidMinClusterEnergy_ (GeV) treshold. Inference is not applied for soft tracksters.
+  // 3. Create input and output tensors. The batch dimension is determined by the number of
+  //    tracksters passing 2.
+  // 4. Fill input tensors with variables of tracksters. Per layer, tracksters are ordered
+  //    descending by energy. Given that tensor data is contiguous in memory, we can use pointer
+  //    arithmetic to fill values, even with batching.
+  // 5. Batched inference.
+  // 6. Assign regressed energy and id probabilities to each trackster.
+
+  // set default values per trackster, determine if the cluster energy threshold is passed,
+  // and store indices of hard tracksters
+  std::vector<int> tracksterIndices;
+  for (int i = 0; i < (int)tracksters.size(); i++) {
+    // set default values (1)
+    tracksters[i].regressed_energy = 0.;
+    for (size_t j = 0; j < tracksters[i].id_probabilities.size(); j++) {
+      tracksters[i].id_probabilities[j] = 0.;
+    }
+
+    // calculate the cluster energy sum (2)
+    // note: after the loop, sumClusterEnergy might be just above the treshold which is enough to
+    // decide whether to run inference for the trackster or not
+    float sumClusterEnergy = 0.;
+    for (size_t cluster = 0; cluster < tracksters[i].vertices.size(); cluster++) {
+      sumClusterEnergy += (float)layerClusters[tracksters[i].vertices[cluster]].energy();
+      // there might be many clusters, so try to stop early
+      if (sumClusterEnergy >= eidMinClusterEnergy_) {
+        tracksterIndices.push_back(i);
+        break;
+      }
+    }
+  }
+
+  // create input and output tensors (3)
+  int batchSize = (int)tracksterIndices.size();
+
+  tensorflow::TensorShape shape({batchSize, eidNLayers_, eidNClusters_, eidNFeatures_});
+  tensorflow::Tensor input(tensorflow::DT_FLOAT, shape);
+  tensorflow::NamedTensorList inputList = {{eidInputName_, input}};
+  float *inputData = input.flat<float>().data();
+
+  std::vector<tensorflow::Tensor> outputs;
+  std::vector<std::string> outputNames;
+  if (!eidOutputNameEnergy_.empty()) {
+    outputNames.push_back(eidOutputNameEnergy_);
+  }
+  if (!eidOutputNameId_.empty()) {
+    outputNames.push_back(eidOutputNameId_);
+  }
+
+  // fill input tensor (4)
+  for (int i : tracksterIndices) {
+    // get features per layer and cluster and store them in a nested vector
+    // this is necessary since layer information is stored per layer cluster, not vice versa
+    // also, this allows for convenient re-ordering
+    std::vector<std::vector<std::array<float, eidNFeatures_> > > tracksterFeatures;
```

Comment:
Repeating comment: memory locality of nested vectors is poor. It would be good to consider better structures.